src/composeappmanager.cc

```diff
@@ -1,23 +1,67 @@
 #include "composeappmanager.h"
 
 #include "boost/process.hpp"
+#include "crypto/crypto.h"
+#include "http/httpclient.h"
+#include "json/json.h"
 
 struct ComposeApp {
-  ComposeApp(std::string name, const ComposeAppConfig &config)
-      : name_(std::move(name)),
-        root_(config.apps_root / name_),
-        compose_(boost::filesystem::canonical(config.compose_bin).string() + " ") {}
+  ComposeApp(std::string name, const ComposeAppConfig& config, const RegistryClient& registry_client)
+      : name_{std::move(name)},
+        root_{config.apps_root / name_},
+        compose_{boost::filesystem::canonical(config.compose_bin).string() + " "},
+        registry_client_{registry_client} {}
+
+  static constexpr const char* const ArchiveExt{".tgz"};
+
+  struct Manifest : Json::Value {
+    static constexpr const char* const Format{"application/vnd.oci.image.manifest.v1+json"};
+    static constexpr const char* const Version{"v1"};
+
+    Manifest(const Json::Value& value = Json::Value()) : Json::Value(value) {
+      auto manifest_version{(*this)["annotations"]["compose-app"].asString()};
+      if (manifest_version.empty()) {
+        throw std::runtime_error("Got invalid App manifest, missing a manifest version: " +
+                                 Utils::jsonToCanonicalStr(value));
+      }
+      if (Version != manifest_version) {
+        throw std::runtime_error("Got unsupported App manifest version: " + Utils::jsonToCanonicalStr(value));
+      }
+    }
+
+    std::string archiveDigest() const {
+      auto digest{(*this)["layers"][0]["digest"].asString()};
+      if (digest.empty()) {
+        throw std::runtime_error("Got invalid App manifest, failed to extract App Archive digest from App manifest: " +
+                                 Utils::jsonToCanonicalStr(*this));
+      }
+      return digest;
+    }
+
+    size_t archiveSize() const {
+      LargestUInt arch_size{(*this)["layers"][0]["size"].asLargestUInt()};
+      if (0 == arch_size || arch_size > std::numeric_limits<size_t>::max()) {
+        throw std::runtime_error("Invalid size of App Archive is specified in App manifest: " +
+                                 Utils::jsonToCanonicalStr(*this));
+      }
+
+      return arch_size;
+    }
+  };
 
   // Utils::shell isn't interactive. The compose commands can take a few
   // seconds to run, so we use boost::process:system to stream it to stdout/sterr
-  bool cmd_streaming(const std::string &cmd) {
+  bool cmd_streaming(const std::string& cmd) {
     LOG_DEBUG << "Running: " << cmd;
     return boost::process::system(cmd, boost::process::start_dir = root_) == 0;
   }
 
-  bool fetch(const std::string &app_uri) {
+  bool download(const std::string& app_uri);
```

Comment:
I'm not against breaking these out. However, it confused my eyes when I first saw this. Maybe you should create a change before this one that breaks out all the methods of ComposeApps to be like this. Then when you do this change it won't look so different from everything else.

Reply:
ok.

Reply:
I just moved it (`download`) to the private methods